### PR TITLE
Fixed #13312 -- Added a way to customize the order_by() of null fields

### PR DESCRIPTION
**Pull request by felixxm**
I prepared new PR based on [6981](https://github.com/django/django/pull/6981) due to inactivity. I also fixed test failure on Oracle and made few small requested changes.

### CODE REVIEW

File: tests/ordering/tests.py [hunk 1]
```diff
@@ -88,6 +88,52 @@ def test_order_by_override(self):
...
+        with self.assertRaisesMessage(ValueError, msg):
+            Article.objects.order_by(F("author").desc(nulls_last=True, nulls_first=True))
+
+    def test_order_by_nulls_last(self):
+        Article.objects.filter(headline="Article 3").update(author=self.author_1)
+        Article.objects.filter(headline="Article 4").update(author=self.author_2)
+        # asc and desc should be chainable with nulls_last
+        self.assertSequenceEqual(
+            Article.objects.order_by(F("author").desc(nulls_last=True)), [
+                self.a4,
```
Comment (timgraham) [inline tests/ordering/tests.py]: I'd single line these to save lines: `[self.a4, ...]`

File: docs/ref/models/expressions.txt [hunk 2]
```diff
@@ -560,13 +560,25 @@ calling the appropriate methods on the wrapped expression.
         nested expressions. ``F()`` objects, in particular, hold a reference
         to a column.
 
-    .. method:: asc()
+    .. method:: asc(**kwargs)
```
Comment (timgraham) [inline docs/ref/models/expressions.txt]: This documentation seems to be in the wrong spot. This is under `Expression` not `F`. Should `BaseExpression` also have these changes or should the documentation be corrected?

Comment (felixxm) [inline docs/ref/models/expressions.txt]: I added changes to `BaseExpression` and updated tests.

**Comment by felixxm** [comment]
Buildbot, test on Oracle.

**Comment by felixxm** [comment]
@timgraham test on Oracle please.

**Comment by timgraham** [comment]
I think you should have permission to run the tests on your own PRs. Did you try "buildbot, test on oracle." as per [Jenkins](https://code.djangoproject.com/wiki/Jenkins)?

**Comment by felixxm** [comment]
buildbot, test on oracle

**Comment by timgraham** [comment]
(trailing period required)

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by timgraham** [comment]
I updated the job configuration, please try once more.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
Buildbot, test on Oracle.

**Comment by felixxm** [comment]
Buildbot, test on Oracle.

File: django/db/models/expressions.py [hunk 3]
```diff
@@ -894,6 +903,22 @@ def as_sql(self, compiler, connection, template=None, **extra_context):
...
+        if self.nulls_last:
+            template = '%(expression)s IS NULL, %(expression)s %(ordering)s'
+        elif self.nulls_first:
+            template = '%(expression)s IS NOT NULL, %(expression)s %(ordering)s'
+        return self.as_sql(compiler, connection, template=template)
+
+    def as_mysql(self, compiler, connection):
+        template = None
+        if self.nulls_last:
+            template = 'IF(ISNULL(%(expression)s),1,0),%(expression)s %(ordering)s '
```
Comment (charettes) [inline django/db/models/expressions.py]: Add a space after the `IF(...),` comma?

File: django/db/models/expressions.py [hunk 4]
```diff
@@ -894,6 +903,22 @@ def as_sql(self, compiler, connection, template=None, **extra_context):
...
+        elif self.nulls_first:
+            template = 'IF(ISNULL(%(expression)s),0,1),%(expression)s %(ordering)s '
```
Comment (charettes) [inline django/db/models/expressions.py]: Ditto.

File: docs/ref/models/expressions.txt [hunk 5]
```diff
@@ -79,6 +79,26 @@ Built-in Expressions
...
+    .. method:: desc(nulls_first=False, nulls_last=False)
+
+        Returns the expression ready to be sorted in descending order.
+
+        ``nulls_first`` and ``nulls_last`` define how null values are sorted.
+
+        .. versionchanged:: 1.11
+
+            The ``nulls_first`` and ``nulls_last`` parameters were added.
+
```
Comment (charettes) [inline docs/ref/models/expressions.txt]: I'm not sure this is worth adding as `asc()` and `desc()` were not documented as `F()` should be considered an expression?

File: docs/releases/1.11.txt [hunk 6]
```diff
@@ -337,6 +337,9 @@ Models
 * You can now use the ``unique=True`` option with
   :class:`~django.db.models.FileField`.
 
+* Added the ``nulls_first`` and ``nulls_last`` parameters to
+  :class:`django.db.models.F` ``asc`` and ``desc`` methods.
+
```
Comment (charettes) [inline docs/releases/1.11.txt]: We should mention expression got support for that too!

**Comment by felixxm** [comment]
Buildbot, test on Oracle.

File: docs/ref/models/expressions.txt (hunk 5 above)
Comment (felixxm) [inline docs/ref/models/expressions.txt]: You're right that `asc` and `desc` were not documented for `F`, but maybe it's worth to mention that something was changed.

**Comment by felixxm** [comment]
I updated PR. Thanks for review.

**Pull request merged**